### ISSUE TITLE
Documentation omission: compactSize encodings must be canonical

### ISSUE THREAD

**Issue by daira** [bitcoin/bitcoin#8721]
The [documentation of a compactSize field](https://bitcoin.org/en/developer-reference#compactsize-unsigned-integers) in the Bitcoin Developer Reference describes four possible encodings based on length. The ranges of integers expressible in each encoding overlap. Therefore it is possible in principle to represent a compactSize with a non-canonical encoding that uses more than the minimum number of bytes.

The actual implementation of [ReadCompactSize](https://github.com/bitcoin/bitcoin/blob/master/src/serialize.h#L274) excludes this possibility. This is also [tested](https://github.com/bitcoin/bitcoin/blob/master/src/test/serialize_tests.cpp#L215). However, it is not documented. Since it affects consensus, it should be documented.

(Ref: https://github.com/zcash/zcash/issues/842 , which is a potential security bug that the Zcash fork of Bitcoin would have had if canonicity were not enforced.)

**Comment by TheBlueMatt** [comment]
I believe back when the documentation was written, we did not enforce compact size, however we did re-serialize blocks (and I believe transactions) when using compactsize for consensus-critical things (eg serialized size checks). For simplicity and to make correctness more obvious, we now enforce the minimal encoding, but I not believe "it affects consensus" aside from being a rule on the p2p network.
Still, this should be filed at https://github.com/bitcoin-dot-org/bitcoin.org/ as that is where the documentation in question lives.
